### PR TITLE
Replace DotDict with Box (except Context)

### PR DESCRIPTION
**Pull request by jlowin**
**Thanks for contributing to Prefect!**

Please describe your work and make sure your PR:

- [x] adds new tests (if appropriate)
- [x] updates `CHANGELOG.md` (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

Note that your PR will not be reviewed unless all three boxes are checked.

## What does this PR change?
This is a large PR but one most people should never notice. It replaces our homegrown `DotDict` with `Box` objects from https://github.com/cdgriffith/Box. Box. For convenience, I left the `Config` and `GraphQLResult` classes in place, but they are simply Box passthrough subclasses.

Principal advantages of adopting Box:
- pays down debt of maintaining our own DotDict class and many subclasses
- `dict` subclass, which (we learned a year ago) is incredibly hard to do right, but extremely valuable because `Box` objects need no special handling
- has optional `defaultdict` semantics that can replace Prefect's `Config` object's `get_nested`, `set_nested`, and `setdefault_nested` methods. In this PR, these semantics are used for building nested license objects trivially.
- has optional `frozen` semantics that aren't used in this PR but could easily be used to enforce immutable objects

Principal disadvantages of adopting Box:
- none
- Ok, I found one odd situation when we pop `serialized_state` out of a deeply nested GraphQLResult. Because Box does some lineage tracking, something about the pop prior to fully "introspecting" the box object caused something in the Box internals to be out of sync. Interestingly, calling `repr(box)` solves this (which is why I had enormous difficulty debugging the issue). I am having trouble replicating it outside our setup, but if I can I'll open an issue on the Box repo. 

Update: Fixed in https://github.com/cdgriffith/Box/pull/103/

Principally, this means every DotDict reference is almost drop-in replaced with `box.Box`, and every `as_nested_dict` is replaced by simply calling `dict(Box)`.

## Context
I did NOT touch the Context object, and therefore `DotDict` and `as_nested_dict` remain in the library (and tested). Context is so central to our machinery that ensuring this switch is handled well is important and deserving of its own PR. I believe that because `Box` is a dict subclass, it can't be trivially combined with `threadlocal` -- however I think we could use "frozen" boxes to create immutable objects that satisfy the same isolated objective.

### CODE REVIEW

File: src/prefect/client/client.py [hunk 1]
```diff
@@ -682,6 +681,13 @@ def get_flow_run_info(self, flow_run_id: str) -> FlowRunInfoResult:
             }
         }
         result = self.graphql(query).data.flow_run_by_pk  # type: ignore
+
+        # this step forces all Box properties to be visited and initialized, which
+        # avoids an odd situation in which the popped serialized_states attempt to
+        # reference box properties that no longer exist. To be clear, this probably
+        # shouldn't be necessary but works.
+        repr(result)
```
Comment (jlowin) [inline src/prefect/client/client.py]: https://github.com/cdgriffith/Box/pull/103

**Comment by jlowin** [comment]
Tests will fail until Box 3.4.4 is released

File: CHANGELOG.md [hunk 2]
```diff
@@ -46,6 +46,7 @@ Released September 20, 2019
 - Allow for Agents to correctly run in environments with differently calibrated clocks - [#1402](https://github.com/PrefectHQ/prefect/issues/1402)
 - Refactor `RemoteEnvironment` to utilize the `get_flow` storage interface - [#1476](https://github.com/PrefectHQ/prefect/issues/1476)
 - Ensure Task logger is available in context throughout every pipeline step of the run - [#1509](https://github.com/PrefectHQ/prefect/issues/1509)
+- Replace `DotDict` with `box.Box` - [#1518](https://github.com/PrefectHQ/prefect/pull/1518)
```
Comment (cicdw) [inline CHANGELOG.md]: @jlowin you'll need to move this to the appropriate place in the changelog now that we've cut a release

File: src/prefect/client/secrets.py [hunk 3]
```diff
@@ -107,5 +106,5 @@ def get(self) -> Optional[Any]:
                 }
                 """,
                 variables=dict(name=self.name),
-            )  # type: Any
-            return as_nested_dict(result.data.secretValue, dict)
+            )
+            return result.data.secretValue
```
Comment (cicdw) [inline src/prefect/client/secrets.py]: Not sure if we have a test for this, but could we make sure that we're testing that `secretValue`, if a pure `dict`, remains a pure `dict`? (i.e., I don't want my secret dictionary to get autoconverted to a Box object)

**Comment by jlowin** [comment]
@cicdw both changes made

**Pull request merged**